File: core/server/project/shine/src/main/java/com/home/shine/support/pool/SafeObjectPool.java
package com.home.shine.support.pool;

import com.home.shine.global.ShineSetting;
import com.home.shine.support.func.ObjectFunc;

/** 对象池(线程安全) */
public class SafeObjectPool<T> extends ObjectPool<T>
{
	public SafeObjectPool(ObjectFunc createFunc)
	{
		super(createFunc);
	}
	
	public SafeObjectPool(ObjectFunc createFunc,int size)
	{
		super(createFunc,size);
	}
	
	@Override
	public T getOne()
	{
		synchronized(this)
		{
			return super.getOne();
		}
	}
	
	public void back(T obj)
	{
		if(!ShineSetting.useObjectPool)
		{
			return;
		}
		
		synchronized(this)
		{
			super.back(obj);
		}
	}
}

File: core/client/game/src/shine/support/pool/SafeObjectPool.cs
using System;

namespace ShineEngine
{
	/// <summary>
	/// 对象池(线程安全)
	/// </summary>
	public class SafeObjectPool<T>:ObjectPool<T>
	{
		public SafeObjectPool(Func<T> createFunc):base(createFunc)
		{
		}

		public SafeObjectPool(Func<T> createFunc,int size):base(createFunc,size)
		{
		}

		public override T getOne()
		{
			lock(this)
			{
				return base.getOne();
			}
		}

		public override void back(T obj)
		{
			lock(this)
			{
				base.back(obj);
			}
		}
	}
}